### PR TITLE
Bootstrap DragonFlyBSD support

### PR DESCRIPTION
**Pull request by driusan**
Add preliminary DragonFlyBSD support to ltsmaster. This is largely based
on copy/pasting the conditional compilation for FreeBSD, but changing
the guard from FreeBSD to DragonFlyBSD.

The only changes from FreeBSD are the pthread_cleanup_pop and
pthread_cleanup_push function names are different in pthread.d, and the
fstat structure is slightly different.

### CODE REVIEW

**Comment by PetarKirov** [comment]
@driusan The general workflow is to first submit pull request(s) upstream (see e.g. the past pull requests for [NetBSD](https://github.com/dlang/druntime/pulls?q=is%3Apr+author%3AnrTQgc) support there), which once merged will be cherry-picked in the ltsmaster branch. The remaining parts can be ironed by opening PRs to ldc-developers/ldc (see for example https://github.com/ldc-developers/ldc/pulls?utf8=✓&q=is%3Apr%20netbsd).

**Comment by driusan** [comment]
@ZombineDev Thanks, but I don't understand how that works for adding support for a platform that isn't currently supported.

Upstream requires a D compiler to compile, and the wiki (https://wiki.dlang.org/Building_LDC_from_source) says that the ltsmaster branch is the latest can be built without access to a D compiler.

(I'm using this to work on adding support upstream, which would presumably get cherry-picked back to the non-lts master. How would I go about building upstream without doing this first?)

**Comment by kinke** [comment]
First of all, thanks for contributing to extending the platform availability for D, much appreciated.

Getting this merged upstream and then cherry-picking it for LDC's master and ltsmaster would be preferrable. That way, the druntime folks can have a discussion about it, and may come up with a neater solution than further copy&paste for all these similar BSD derivatives (4k additional lines for minimal actual differences is a bit much IMO).

**Comment by driusan** [comment]
I've been trying to use this to add support upstream (the version I have for upstream is able to compile dmd using this, and the compiled dmd is able to compile, but not link the result, so I'm more confident in this ltsmaster version than I am with what I have there..)

I assumed that that the workflow was ltsmaster->upstream->master, but I'll close this and just send upstream once it's working.

**Comment by redstar** [comment]
@driusan Thanks for your work!!! I really appreciate to see more *BSD support!

There are a lot of version in upstream Phobos/druntime which are not supported by dmd. The main reason for requesting adding the PR to upstream first is minimizing the difference between the upstream and the LDC version of Phobos/druntime. It is much, much easier to maintain it in this way.
As a byproduct, the GDC people have a chance to pick up the new code, too.

**Comment by JohanEngelen** [comment]
@driusan I feel for your ltsmaster->upstream approach, so that you can actually test the code before it is committed. Perhaps explicitly asking some of the druntime folks to look at some specific ltsmaster PR is a good way to go. (so that the cherry-pick onto upstream is trivial)

File: src/core/sys/dragonflybsd/sys/event.d [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    EVFILT_AIO      =  -3, /* attached to aio requests */
+    EVFILT_VNODE    =  -4, /* attached to vnodes */
+    EVFILT_PROC     =  -5, /* attached to struct proc */
+    EVFILT_SIGNAL   =  -6, /* attached to struct proc */
+    EVFILT_TIMER    =  -7, /* timers */
+    // EVFILT_NETDEV   =  -8, /* no longer supported */
+    EVFILT_FS       =  -9, /* filesystem events */
+    EVFILT_LIO      = -10, /* attached to lio requests */
+    EVFILT_USER     = -11, /* User events */
+    EVFILT_SYSCOUNT =  11,
```
Comment (ivadasz) [inline src/core/sys/dragonflybsd/sys/event.d:35]: These EVFILT_* values don't match the DragonFlyBSD headers.

File: src/core/sys/dragonflybsd/sys/event.d [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    EV_CLEAR    = 0x0020, /* clear event state after reporting */
+    EV_RECEIPT  = 0x0040, /* force EV_ERROR on success, data=0 */
+    EV_DISPATCH = 0x0080, /* disable event after reporting */
+
+    EV_SYSFLAGS = 0xF000, /* reserved by system */
+    EV_FLAG1    = 0x2000, /* filter-specific flag */
+
+    /* returned values */
+    EV_EOF      = 0x8000, /* EOF detected */
+    EV_ERROR    = 0x4000, /* error, data contains errno */
```
Comment (ivadasz) [inline src/core/sys/dragonflybsd/sys/event.d:72]: DragonFlyBSD has an additional EV_NODATA enum value here.

File: src/core/sys/dragonflybsd/sys/event.d [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    NOTE_FFLAGSMASK = 0x00ffffff,
+
+    NOTE_TRIGGER    = 0x01000000, /* Cause the event to be
+                                  triggered for output. */
+
+    /*
+     * data/hint flags for EVFILT_{READ|WRITE}, shared with userspace
+     */
+    NOTE_LOWAT      = 0x0001, /* low water mark */
+
```
Comment (ivadasz) [inline src/core/sys/dragonflybsd/sys/event.d:100]: a NOTE_OOB enum value exists in DragonFlyBSD. It seems like a good idea to compare this whole file against the sys/event.h header in DragonFlyBSD.
Note that DragonFlyBSD doesn't have as long support periods as FreeBSD, so it should be easy to include more modern kqueue arguments here, than the FreeBSD might have.

**Comment by ivadasz** [comment]
There seem to be a lot of cases in the diff, where either enum values don't match the dragonflybsd headers, or various flags that can be quite useful are missing.
I would suggest taking maybe DragonFlyBSD 4.8 as a baseline for support, but not older versions like 4.6.

**Comment by driusan** [comment]
thanks for pointing those out, @ivadasz! I'm still working on trying to make a version of this PR for upstream (but haven't had much time to work on it lately), do you have any pointers for system headers that commonly differ between DragonFly and FreeBSD that I should keep an eye on other than event.(d/h)?

I did a spot check before sending this PR, but I obviously missed some..

**Comment by PetarKirov** [comment]
@driusan I'm afraid I may have put you off the right track. I didn't mean to say that you should develop full DragonFlyBSD  support with dmd before it is merged into ldc. On the contrary, you should take approach that is most convenient for you for the initial "bring up" ( as hardware devs use to say). This may easily be ldc's lts-master branch, since in general it attracts much more attention than whatever dmd has as the last supported C++ version (pretty much nothing).
Developing the support for new platforms on LDC or GDC is natural and expected as the focus of dmd is developing language features and testing them on a limited number of well supported platforms. Because of that dmd's bootstrapping process receives close to no attention (as it is "good enough" for the platforms we have CI testing on).

What I meant to say is that once you feel ready with the DragonFlyBSD support (perhaps developed on ldc lts-master, or whatever works for you) you should post your druntime and/or phobos  specific changes upstream, so the a bit broader community can review them. Even when the changes are merged, the dmd support may still not be ready for whatever reasons (e.g. platform X has a different ABI which LLVM natively supports, and for the moment you're not interesting in touching dmd's backend) and this ok.
Some of the reasons behind merging code upstream first are:
1) makes the development easier to keep track of - when you say D version 2.XXX you know for sure exactly what platform support is there there in druntime/phobos regardless of the compiler - dmd, gdc or ldc)
2) when a pull request is posted upstream more people get a chance to review it
3) once preliminary support lands upstream it is much easier for it to reach all the other compilers

**Pull request closed without merging**